qcodes/dataset/data_set.py

```diff
@@ -818,18 +875,30 @@ def add_result_to_queue(self,
         expected_keys = frozenset.union(*[frozenset(d) for d in results])
         values = [[d.get(k, None) for k in expected_keys] for d in results]
 
-        item = {'keys': list(expected_keys), 'values': values}
-        self._data_write_queue.put(item)
+        item = {'keys': list(expected_keys), 'values': values,
+                "table_name": self.table_name}
+        writer_status = self._writer_status
+        if writer_status is None:
+            raise ValueError("Dataset writer has not been setup")
+        writer_status.data_write_queue.put(item)
+
+    def _ensure_dataset_written(self) -> None:
+        writer_status = self._writer_status
+        if writer_status is None:
+            raise ValueError("Dataset writer has not been setup")
+
+        if writer_status.write_in_background:
+            writer_status.data_write_queue.put({'keys': 'finalize', 'values': self.run_id})
+            while self.run_id in writer_status.active_datasets:
+                time.sleep(1e-3)
```

Comment:
One could replace this with a threading.event to signal the dataset writting has been complted

Reply:
would that be better because one doesn't need to define the time.sleep? or because the code would be more straightforward to read?

i am personally ok with the sleep here

Reply:
Mostly since the sleep time is a bit random. If its too long this will potentially be very slow if too small we are wasting cpu cycles in this busy loop but the current value seems to work well enough

Reply:
ok. then it's fine for now, i guess

Reply:
the sleeping time can be made a attribute/class-variable ;)

Reply:
Sure why not.